File: internal/bridge/ibmmq.go
package bridge

import (
	"context"

	ibmmqdriver "github.com/amigoer/mq-studio/internal/driver/ibmmq"
	"github.com/amigoer/mq-studio/internal/model"
	ibmmqservice "github.com/amigoer/mq-studio/internal/service/ibmmq"
)

// IBMMQService is the renderer's entry point for the operations only IBM MQ
// has. Listing and describing queues and topics go through the canonical
// services; what is here is the rest.
type IBMMQService struct {
	service *ibmmqservice.Service
}

// NewIBMMQService wires the bridge to the service.
func NewIBMMQService(service *ibmmqservice.Service) *IBMMQService {
	return &IBMMQService{service: service}
}

// QueueManager is which queue manager this connection speaks to.
func (s *IBMMQService) QueueManager(connID int) (string, error) {
	return s.service.QueueManager(connID)
}

// IBMMQDestinationInput is a queue or a topic as the form collects it.
//
// Deliberately not TopicService.Create's shape. That one takes a broker
// address, two queue counts and a permission string - RocketMQ's vocabulary,
// of which neither an MQ queue nor an MQ topic has any.
type IBMMQDestinationInput struct {
	Name string `json:"name"`
	// Kind is "queue" or "topic", and it decides which interface the create
	// goes through: a queue is a REST resource and a topic is MQSC.
	Kind string `json:"kind"`

	// QueueType is local, alias, remote or model. Only a local queue stores
	// anything; the rest resolve somewhere else.
	QueueType string `json:"queueType"`
	// MaxDepth caps how many messages a local queue will hold. Zero leaves the
	// queue manager's own default, which is 5000 on a fresh installation.
	MaxDepth int `json:"maxDepth"`

	// TopicString is what publishers name, and it is required on a topic. It
	// is not the object's name: the object is where that string's settings are
	// attached, and two objects covering overlapping strings is ordinary.
	TopicString string `json:"topicString"`

	Description string `json:"description"`
}

func (input IBMMQDestinationInput) spec() model.DestinationSpec {
	attributes := map[string]string{
		ibmmqdriver.AttrKind:        input.Kind,
		ibmmqdriver.AttrQueueType:   input.QueueType,
		ibmmqdriver.AttrTopicString: input.TopicString,
		ibmmqdriver.AttrDescription: input.Description,
	}
	if input.MaxDepth > 0 {
		attributes[ibmmqdriver.AttrMaxDepth] = itoa(input.MaxDepth)
	}
	return model.DestinationSpec{
		Ref:        model.DestinationRef{Name: input.Name},
		Attributes: attributes,
	}
}

// CreateDestination declares a queue or a topic on the queue manager.
func (s *IBMMQService) CreateDestination(connID int, input IBMMQDestinationInput) error {
	return s.service.CreateDestination(context.Background(), connID, input.spec())
}

// RemoveDestination deletes a queue or a topic.
//
// purge decides what happens to a queue that is not empty: without it the
// queue manager refuses, with it the messages go too and there is no undo.
// A queue an application has open is refused either way.
func (s *IBMMQService) RemoveDestination(connID int, name string, purge bool) error {
	return s.service.RemoveDestination(context.Background(), connID, name, purge)
}

File: frontend/src/api/ibmmq.ts
import { IBMMQService } from "@bindings/bridge";
import type {
  IBMMQDestinationInput,
  IBMMQPublishInput,
  IBMMQPublishResult,
} from "@bindings/bridge/models";
import type { Channel, DeadLetterQueue } from "@bindings/model/models";
import { present, required } from "./client";

export type { IBMMQDestinationInput, IBMMQPublishInput, IBMMQPublishResult };

/**
 * The IBM MQ-only half of the surface.
 *
 * Reading queues and topics is not here: both are destinations, and
 * api/topic.ts already answers the whole read side. What this carries is what
 * the canonical services cannot express - a create whose fields depend on
 * which of the two objects is being made, where TopicService.Create collects a
 * broker address, two queue counts and a permission string.
 */

/** Which queue manager this connection speaks to. */
export const queueManager = (connID: number): Promise<string> => IBMMQService.QueueManager(connID);

/**
 * Declare a queue or a topic on the queue manager.
 *
 * The kind decides everything else: a queue takes a type and a maximum depth,
 * a topic takes the string publishers name. A topic without one would be an
 * object nothing ever publishes through.
 */
export const createDestination = (connID: number, input: IBMMQDestinationInput): Promise<void> =>
  IBMMQService.CreateDestination(connID, input);

/**
 * Delete a queue or a topic.
 *
 * Without purge the queue manager refuses a queue that holds messages, which
 * is the check worth keeping; with it the messages go too and there is no
 * undo. A queue an application has open is refused either way.
 */
export const removeDestination = (connID: number, name: string, purge: boolean): Promise<void> =>
  IBMMQService.RemoveDestination(connID, name, purge);

/**
 * Every channel definition, with what is running folded in.
 *
 * Definitions rather than connections: a channel is listed whether or not
 * anything is using it, because it is what decides whether anything may.
 */
export const channels = async (connID: number): Promise<Channel[]> =>
  present(await IBMMQService.Channels(connID));

/**
 * Send one body, or the same body several times, to one queue.
 *
 * A queue, not a topic: the messaging REST API has no topic resource, so
 * publishing needs an MQ client and this console does not pretend otherwise.
 * Each copy is its own request, which is why the count is capped.
 */
export const publish = (
  connID: number,
  input: IBMMQPublishInput,
): Promise<IBMMQPublishResult> => IBMMQService.Publish(connID, input).then(required);

/**
 * The queues something else dead-letters into.
 *
 * Found by walking every queue's configuration backwards: nothing marks a
 * dead-letter queue here, and what makes one is the queue manager's DEADQ
 * attribute or another queue's backout queue pointing at it.
 */
export const deadLetterQueues = async (connID: number): Promise<DeadLetterQueue[]> =>
  present(await IBMMQService.DeadLetterQueues(connID));
